### PR TITLE
MDEV-15364 FOREIGN CASCADE operations in system versioned referenced tables

### PR DESCRIPTION
**Pull request by kevgs**
I'm contributing this new code of the whole pull request, including one or several files that are either new files or modified ones, under the BSD-new license.

### CODE REVIEW

**Comment by kevgs** [comment]
[Buildbot](http://buildbot.askmonty.org/buildbot/grid?category=main&branch=tt-10.3-kevgs)

**Comment by midenok** [comment]
Duplicate of #668.

File: storage/innobase/include/data0type.h [hunk 1]
```diff
@@ -543,18 +542,21 @@ struct dtype_t{
 					in bytes */
 
 	/** @return whether this is system field */
-	bool vers_sys_field() const { return prtype & DATA_VERSIONED; }
+	bool vers_sys_field() const
+	{
+		return vers_sys_start() || vers_sys_end();
+	}
```
Comment (dr-m) [inline storage/innobase/include/data0type.h]: Why was this changed? I would not count on the compilers to convert the changed code to a single bitwise AND instruction, similar to how the existing code was probably translated.

File: storage/innobase/include/dict0mem.h [hunk 2]
```diff
@@ -653,18 +653,28 @@ struct dict_col_t{
...
+	{
+		return vers_sys_start() || vers_sys_end();
+	}
+	/** @return whether table of this system field is TRX_ID-based */
+	bool vers_native() const
+	{
+		ut_ad(vers_sys_field());
+		ut_ad(mtype == DATA_INT || mtype == DATA_FIXBINARY);
+		return mtype == DATA_INT;
+	}
```
Comment (dr-m) [inline storage/innobase/include/dict0mem.h]: I think that it is OK to add the new member function `vers_native()`, but I oppose the change to `vers_sys_field()`.

File: storage/innobase/include/row0upd.h [hunk 3]
```diff
@@ -577,6 +582,9 @@ struct upd_node_t{
 				/* column assignment list */
 	ulint		magic_n;
 
+	/** System Versioning: modify update vector to set row_start
+	 * (or row_end in case of DELETE) to current trx_id. */
+	void		vers_set_fields(const trx_t* trx);
```
Comment (dr-m) [inline storage/innobase/include/row0upd.h]: When and how exactly is the update vector modified? Are new fields added? Do not add `*` in the middle of Doxygen comments. Replace `trx_id` with `trx->id`. Maybe the function comment should refer to `field->col->vers_native()` or `dict_col_t::vers_native()`?

File: storage/innobase/row/row0mysql.cc [hunk 4]
```diff
@@ -1781,6 +1790,50 @@ init_fts_doc_id_for_ref(
...
+		ufield = &update->fields[0];
+		update->n_fields = 0;
+		is_delete = VERSIONED_DELETE;
+		col_idx = table->vers_end;
+	} else {
+		ut_ad(update->n_fields < table->n_cols);
+		ufield = &update->fields[update->n_fields];
+		col_idx = table->vers_start;
+	}
+	col = &table->cols[col_idx];
```
Comment (dr-m) [inline storage/innobase/row/row0mysql.cc]: Declare `const dict_col_t* col` here where it is assigned.
[C++ Core Guidelines ES.5: Keep scopes small](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-scope)

File: storage/innobase/row/row0mysql.cc [hunk 5]
```diff
@@ -1781,6 +1790,50 @@ init_fts_doc_id_for_ref(
...
+		ut_ad(field_no != ULINT_UNDEFINED);
+		ufield->field_no = field_no;
+	}
+	ufield->orig_len = 0;
+	ufield->exp = NULL;
+
+	if (col->vers_native())
+	{
+		mach_write_to_8(update->vers_sys_value, trx->id);
+	} else {
```
Comment (dr-m) [inline storage/innobase/row/row0mysql.cc]: In InnoDB, do not break the line before `{` in `if` or `else` statements.

File: storage/innobase/row/row0mysql.cc [hunk 6]
```diff
@@ -1781,6 +1790,50 @@ init_fts_doc_id_for_ref(
...
+		thd_get_query_start_data(trx->mysql_thd, (char *)
+					update->vers_sys_value);
```
Comment (dr-m) [inline storage/innobase/row/row0mysql.cc]: Do not use C-style casts.
Use `reinterpret_cast<char*>`. Why can’t `vers_sys_value` not be declared in the matching type in the first place?

File: storage/innobase/row/row0mysql.cc [hunk 7]
```diff
@@ -1781,6 +1790,50 @@ init_fts_doc_id_for_ref(
...
+	}
+
+	dfield_t* dfield = &ufield->new_val;
+	dfield_set_data(dfield, update->vers_sys_value, col->len);
+	dict_col_copy_type(col, &dfield->type);
+
+	update->n_fields++;
+	ut_ad(in_mysql_interface); // otherwise needs to recalculate node->cmpl_info
```
Comment (dr-m) [inline storage/innobase/row/row0mysql.cc]: Assertions should be at the start of the function.

File: storage/innobase/row/row0mysql.cc [hunk 8]
```diff
@@ -2174,6 +2192,90 @@ row_mysql_unfreeze_data_dictionary(
...
+/** Function restores btr_pcur_t, creates dtuple_t from rec_t,
+sets row_end = CURRENT_TIMESTAMP/trx->id, inserts it to a table and updates
+table statistics.
+This is used in UPDATE CASCADE/SET NULL of a system versioning table.
+@param[in]	thr	current query thread
+@param[in]	node	a node which just updated a row in a foreign table
+@return DB_SUCCESS or some error */
+UNIV_INLINE
+dberr_t
+row_update_versioned_insert(que_thr_t* thr, const upd_node_t* node)
```
Comment (dr-m) [inline storage/innobase/row/row0mysql.cc]: Use `inline` instead of `UNIV_INLINE`. If the whole signature fits in 79 columns or less, then do not break lines:
```c++
inline dberr_t row_update_versioned_insert(que_thr_t* thr, const upd_node_t* node)
```
It does not fit, but if we abbreviated `update` to `upd` or `versioned` to `vers`, then it would fit.

File: storage/innobase/row/row0mysql.cc [hunk 9]
```diff
@@ -2174,6 +2192,90 @@ row_mysql_unfreeze_data_dictionary(
...
+	} else {
+		mode = BTR_MODIFY_LEAF;
+	}
+
+	bool success = btr_pcur_restore_position(mode, node->pcur, &mtr);
+	mtr.commit();
+
+	if (!success) {
+		return DB_RECORD_NOT_FOUND;
+	}
```
Comment (dr-m) [inline storage/innobase/row/row0mysql.cc]: Why do you think it is OK to `mtr.commit()` after restoring the cursor position?
What is the purpose of restoring the cursor position if you are committing the mini-transaction right after?
Between the `mtr.commit()` and the subsequent mini-transactions that would modify the tree, the state can be changed by other threads. So, the value of `success` is stale.

This check should be done somewhere inside `row_ins_step()` instead. Why not just handle a duplicate key error?

File: storage/innobase/row/row0mysql.cc [hunk 10]
```diff
@@ -2193,51 +2295,23 @@ row_update_cascade_for_mysql(
...
 
-	trx_t* trx = thr_get_trx(thr);
+	const trx_t* trx = thr_get_trx(thr);
 
-	bool vers_set_fields = node->table->versioned()
-				  && (node->is_delete == PLAIN_DELETE
-				      || node->update->affects_versioned());
+	bool vers_set_fields = table->versioned()
+				&& (node->is_delete == PLAIN_DELETE
+				    || node->update->affects_versioned());
```
Comment (dr-m) [inline storage/innobase/row/row0mysql.cc]: The indentation is still too deep. The second line should be indented by one extra TAB.
The relative indentation of the 3rd line to the 2nd line looks correct.

**Comment by kevgs** [comment]
> I would like to keep the current name of DATA_VERSIONED. The DATA_UNVERSIONED is counter-intuitive, because the field would be 0 and not nonzero for unversioned columns.

1) I think its more natural to mark (add flags for) uncommon case. `sql` does so, for example.

2) When we have two tables, one of them system-versioning and they have a `FOREIGN` relationship at some point of code we compare types of fields. And they will be different because one of them is `DATA_VERSIONED`. This requires a special handling. It's more that just renaming.

File: storage/innobase/dict/dict0mem.cc [hunk 11]
```diff
@@ -312,14 +312,15 @@ dict_mem_table_add_col(
...
 
-	switch (prtype & DATA_VERSIONED) {
-	case DATA_VERS_START:
-		ut_ad(!table->vers_start);
-		table->vers_start = i;
-		break;
-	case DATA_VERS_END:
-		ut_ad(!table->vers_end);
-		table->vers_end = i;
+	if ((prtype & DATA_UNVERSIONED) != DATA_UNVERSIONED) {
```
[the remark this answers is not in the document]
Comment (kevgs) [reply storage/innobase/dict/dict0mem.cc]: > I think it would be better to avoid == or != in bitwise operations

I've tried to make code as easy to read as possible. I haven't checked whether current compiles optimize it or not. I'm sure they will in future. And I'm optimize code for reading, not for (possible) current compiles limitations.

`!(~prtype & DATA_UNVERSIONED)` will become `true` for unversioned fields. We're interested here only in `row_start` and `row_end`.

> Why was it necessary to change this code?

Looks like a general question. I answered it somewhere else. It wasn't strictly necessary. But IMO it makes code more clear.

File: storage/innobase/include/data0type.h (hunk 1 above)
Comment (kevgs) [reply storage/innobase/include/data0type.h]: Function returned `true` for `row_start`, `row_end` and every ordinary versioned fields. While `sys_field` means just `row_start` and `row_end`. It was incorrect in that sense. I fixed it with readable, instead of hand-optimized code.

File: storage/innobase/row/row0mysql.cc (hunk 8 above)
Comment (kevgs) [reply storage/innobase/row/row0mysql.cc]: Ok. I've tried to avoid words shortening because of this rule http://google.github.io/styleguide/cppguide#General_Naming_Rules

File: storage/innobase/include/data0type.h (hunk 1 above)
Comment (dr-m) [reply storage/innobase/include/data0type.h]: I think that low-level code should be optimized.
We could surely add `ut_ad()` for documentation or readability, like this:
```c++
bool vers_sys_field() const
{
	bool vers_sys = prtype & DATA_VERSIONED;
	ut_ad(vers_sys == vers_sys_start() || vers_sys_end());
	return vers_sys;
}
```

File: storage/innobase/dict/dict0mem.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (dr-m) [reply storage/innobase/dict/dict0mem.cc]: Sorry, I should have suggested
```c
if (~prtype & DATA_VERSIONED) {
```
This would hold whenever any of the `DATA_VERSIONED` bits are not set.
I would not rename the bitmask. Its current name is descriptive: it is related to versioning. Naming the bitmask for only one value (`DATA_UNVERSIONED`) would make the code harder to read for me. Usually, when a named flag (or group of flags) is set, the name of the bitmask should describe the situation. In this case, `DATA_VERSIONED` is an appropriate name, because the column would have to do something with system versioning.

And yes, there is a bad precedent: Setting `BTR_EXTERN_OWNER_FLAG` actually means ‘disowned’.

File: storage/innobase/row/row0mysql.cc (hunk 9 above)
Comment (kevgs) [reply storage/innobase/row/row0mysql.cc]: What I'm trying to do at this point is to get an old values for updated row to compose historical row. This is the simplest way I found to do it, thanks to `btr_pcur_t` which points to the right place.

Another way to get and old values I considered was reading undo log right update row was updated.

File: storage/innobase/row/row0mysql.cc (hunk 8 above)
Comment (kevgs) [reply storage/innobase/row/row0mysql.cc]: Btw, `static` is better here. Compiler may inline function if it hasn't `inline` specifier. And it will be easier for compiler to do so if it will know a function has a file scope.

File: storage/innobase/include/data0type.h (hunk 1 above)
Comment (midenok) [reply storage/innobase/include/data0type.h]: There is different semantics when both bits are set. It means: field is _unversioned_ (i.e. "optimized"). The above suggestion will falsely return **true** in such cases.

File: storage/innobase/row/row0mysql.cc (hunk 9 above)
Comment (kevgs) [reply storage/innobase/row/row0mysql.cc]: I guess I understand why record lock is needed here: it's a shared data after all. I need to lock it and `mtr.commit()` after `row_build()`.

But I do not understand how can I move `btr_pcur_restore_position()` inside `row_ins_step()`: I need to compose a row to insert before this call, not inside of it.

I also do not understand how to get and how to use a duplicate key error here.

**Comment by kevgs** [comment]
Read rec is locked here:
```c++
#0  lock_clust_rec_read_check_and_lock_alt (flags=0, block=0x7fffedf71700, rec=0x7fffee48c07d "", index=0x7fffa0155fe8, mode=LOCK_X, gap_mode=1024, thr=0x7fffa0104620) at lock0lock.cc:6572
#1  row_ins_foreign_check_on_constraint (thr=0x7fffa0104620, foreign=0x7fffa007f7e8, pcur=0x7fffedc9b198, entry=0x7fffa0179020, mtr=0x7fffedc9ac80) at row0ins.cc:1242
#2  row_ins_check_foreign_constraint (check_ref=0, foreign=0x7fffa007f7e8, table=0x7fffa0006b78, entry=0x7fffa0179020, thr=0x7fffa0104620) at row0ins.cc:1804
#3  row_upd_check_references_constraints (node=0x7fffa0104328, pcur=0x7fffa0032d50, table=0x7fffa0006b78, index=0x7fffa00362b8, offsets=0x7fffa0163ce0, thr=0x7fffa0104620, mtr=0x7fffedc9bda8) at row0upd.cc:326
#4  row_upd_clust_rec_by_insert (node=0x7fffa0104328, index=0x7fffa00362b8, thr=0x7fffa0104620, referenced=1, foreign=false, mtr=0x7fffedc9bda8) at row0upd.cc:2758
#5  row_upd_clust_step (node=0x7fffa0104328, thr=0x7fffa0104620) at row0upd.cc:3212
#6  row_upd (node=0x7fffa0104328, thr=0x7fffa0104620) at row0upd.cc:3288
#7  row_upd_step (thr=0x7fffa0104620) at row0upd.cc:3432
#8  row_update_for_mysql (prebuilt=0x7fffa0103828) at row0mysql.cc:1887
#9  ha_innobase::update_row (this=0x7fffa0102010, old_row=0x7fffa0157248 "\377\001", new_row=0x7fffa0157240 "\377\v") at ha_innodb.cc:9044
#10 handler::ha_update_row (this=0x7fffa0102010, old_data=0x7fffa0157248 "\377\001", new_data=0x7fffa0157240 "\377\v") at handler.cc:6245
#11 mysql_update (thd=0x7fffa0000ce8, table_list=0x7fffa00140b8, fields=List<Item> with 2 elements, values=List<Item> with 2 elements, conds=0x0, order_num=0, order=0x0, limit=18446744073709551615, handle_duplicates=DUP_ERROR, ignore=false, found_return=0x7fffedc9ed28, updated_return=0x7fffedc9ed20) at sql_update.cc:942
#12 mysql_execute_command (thd=0x7fffa0000ce8) at sql_parse.cc:4565
#13 mysql_parse (thd=0x7fffa0000ce8, rawbuf=0x7fffa0013fb0 "update parent set id = 11, value = value + 1", length=44, parser_state=0x7fffedca0400, is_com_multi=false, is_next_command=false) at sql_parse.cc:8009
```

It is locked for row update and I reuse that lock for my purposes.

Dedicated `mtr_t` is not used anymore and record is read in other place. `upd_node_t::historical_row` was added for this purpose.

Other (mostly style) things was fixed.

File: storage/innobase/include/row0upd.h [hunk 12]
```diff
@@ -568,6 +568,8 @@ struct upd_node_t{
 	dtuple_t*	row;	/*!< NULL, or a copy (also fields copied to
 				heap) of the row to update; this must be reset
 				to NULL after a successful update */
+	dtuple_t*	historical_row;	/*!< historical row used in
+				CASCADE UPDATE/SET NULL */
```
Comment (dr-m) [inline storage/innobase/include/row0upd.h]: The comment should say where the memory for `historical_row` is allocated from, and what its lifetime is (that is, when it is assigned non-`NULL` and back to `NULL`).

File: storage/innobase/row/row0ins.cc [hunk 13]
```diff
@@ -1391,6 +1393,13 @@ row_ins_foreign_check_on_constraint(
 		}
 	}
 
+	if (table->versioned() && cascade->is_delete != PLAIN_DELETE
+	    && cascade->update->affects_versioned()) {
+		cascade->historical_row =
+			row_build(ROW_COPY_DATA, clust_index, clust_rec, NULL,
+				  table, NULL, NULL, NULL, thr->prebuilt->heap);
```
Comment (dr-m) [inline storage/innobase/row/row0ins.cc]: If we allocate this from `row_prebuilt_t::heap`, then the memory consumption will grow until the end of the statement, right?
Do we really need all the `historical_row` so long? What if the statement updated thousands or millions of rows?
And do we need `ROW_COPY_DATA`?

File: storage/innobase/row/row0upd.cc [hunk 14]
```diff
@@ -3454,3 +3454,57 @@ row_upd_step(
...
+	const dict_col_t* col = dict_table_get_nth_col(table, idx);
+
+	upd_field_set_field_no(ufield, dict_col_get_clust_pos(col, clust_index),
+			       clust_index);
+
+	char* where = reinterpret_cast<char*>(update->vers_sys_value);
+	if (col->vers_native()) {
+		mach_write_to_8(where, trx->id);
+	} else {
+		thd_get_query_start_data(trx->mysql_thd, where);
```
Comment (dr-m) [inline storage/innobase/row/row0upd.cc]: Which format is `thd_get_query_start_data()` storing the data in? Big-endian, and (in case of signed integer) inverted sign bit?

Comment (kevgs) [reply storage/innobase/row/row0upd.cc]: First 4 bytes big-endian `int`. And ordinary timestamp.
Next 3 bytes is big-endian 'int_24`. Max value is 999999. Represents microseconds.

Max `DATETIME` example https://github.com/MariaDB/server/blob/10.3/storage/innobase/trx/trx0trx.cc#L67
Btw, it should be called `datetime_max_bytes`.

**Comment by kevgs** [comment]
Memory consumption is fixed by adding a dedicated `mem_heap_t`. Changes are in separate commits. I would squash them to the third commit before merging to `10.3`.
